### PR TITLE
[ticket/17151] Make settings forms use macros

### PR DESCRIPTION
**Pull request by rxu**
PHPBB3-17151

Checklist:

- [x] Correct branch: master for new features; 3.3.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/master/coding-guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

Tracker ticket:

https://tracker.phpbb.com/browse/PHPBB3-17151

### CODE REVIEW

File: tests/functional/fixtures/ext/foo/bar/acp/main_module.php [hunk 1]
```diff
@@ -49,6 +49,7 @@ function main($id, $mode)
 				'setting_7'	=> ['lang' => 'SETTING_7',	'validate' => 'email',		'type' => 'email:0:100', 'explain' => true],
 				'setting_8'	=> ['lang' => 'SETTING_8',	'validate' => 'string',		'type' => 'textarea:5:30', 'explain' => true],
 				'setting_9'	=> ['lang' => 'SETTING_9',	'validate' => 'bool',		'type' => 'radio:enabled_disabled', 'explain' => true],
+				'setting_10'=> ['lang' => 'SETTING_10',	'validate' => 'bool',		'type' => 'radio:{"LABEL_1":1, "LABEL_3":3, "LABEL_2":2}', 'explain' => true],
```
Comment (Derky) [inline tests/functional/fixtures/ext/foo/bar/acp/main_module.php:52]: Could you have a look at how this is currently solved for the <select> form input? Because a list of radio buttons basically results in the same output as a (single) select dropdown.

File: phpBB/includes/functions_acp.php [hunk 2]
```diff
@@ -367,83 +374,42 @@ function phpbb_build_cfg_template(array $tpl_type, string $key, &$new_ary, $conf
...
-
-	$tpl = '';
-	$name = 'config[' . $config_key . ']';
-
-	// Make sure there is no notice printed out for non-existent config options (we simply set them)
-	if (!isset($new_ary[$config_key]))
-	{
-		$new_ary[$config_key] = '';
-	}
+			$params = json_decode($tpl_type[1], true);
```
Comment (Derky) [inline phpBB/includes/functions_acp.php:384]: Could you maybe change the radio options to a similar way on how it's done with generating the select dropdown? So that you can just pass an array of options for the radio buttons?

**Review by Derky — CHANGES_REQUESTED** [review]

File: phpBB/styles/all/template/macros/forms/select.twig [hunk 3]
```diff
@@ -21,11 +21,18 @@
 			label="{{ element.label }}">
 		{% endapply %}
 		{% for option in element.options %}
-			<option value="{{ option.value }}"{% if option.selected %} selected="selected"{% endif %}{% if option.disabled %} disabled="disabled" class="disabled-option"{% endif %}>{{ option.label }}</option>
+			<option
+				value="{{ option.value }}"
+				{% if option.selected %}selected="selected"{% endif %}
+				{% if option.disabled %} disabled="disabled"{% endif %}{% if option.class %} class="{{ option.class }}"{% endif %}>{{ option.label }}</option>
```
[the remark this answers is not in the document]
Comment (rxu) [reply phpBB/styles/all/template/macros/forms/select.twig:27]: That class option works for the whole `select` while this one works on per-option basis.

File: phpBB/includes/functions_acp.php (hunk 2 above)
Comment (rxu) [reply phpBB/includes/functions_acp.php:384]: The problem here is that the options are being set in a string within ACP module class. As for `select`, they're only generated using class method. But if this change is welcome I can do it like that (BC will be broken).

File: phpBB/includes/functions_acp.php [hunk 4]
```diff
@@ -483,38 +449,41 @@ function build_cfg_template($tpl_type, $key, &$new_ary, $config_key, $vars)
...
 					$tpl = [
 						'tag'			=> 'select',
+						'class'			=> $tpl_type['class'] ?? false,
 						'id'			=> $key,
+						'data'			=> $tpl_type['data'] ?? [],
 						'name'			=> $name,
-						'toggleable'	=> !empty($tpl_type[2]),
-						'options'		=> $return,
+						'toggleable'	=> !empty($tpl_type[2]) || !empty($tpl_type['toggleable']),
+						'options'		=> $tpl_type['options'],
```
[the remark this answers is not in the document]
Comment (rxu) [reply phpBB/includes/functions_acp.php:465]: How to reproduce?

[the remark this answers is not in the document]
Comment (rxu) [reply phpBB/includes/functions_acp.php:465]: Oh, indeed. Weird that tests didn't fail on that.

File: phpBB/includes/acp/acp_attachments.php [hunk 5]
```diff
@@ -176,26 +176,26 @@ function main($id, $mode)
 						'img_link_width'		=> array('lang' => 'IMAGE_LINK_SIZE', 'validate' => 'int:0', 'type' => false, 'method' => false, 'explain' => false,),
 						'img_link_height'		=> array('lang' => 'IMAGE_LINK_SIZE', 'validate' => 'int:0', 'type' => false, 'method' => false, 'explain' => false,),
 
-						'allow_attachments'		=> array('lang' => 'ALLOW_ATTACHMENTS',		'validate' => 'bool',	'type' => 'radio:yes_no', 'explain' => false),
-						'allow_pm_attach'		=> array('lang' => 'ALLOW_PM_ATTACHMENTS',	'validate' => 'bool',	'type' => 'radio:yes_no', 'explain' => false),
+						'allow_attachments'		=> array('lang' => 'ALLOW_ATTACHMENTS',		'validate' => 'bool',	'type' => 'radio', 'function' => 'build_radio', 'params' => ['{CONFIG_VALUE}', '{KEY}', [1 => 'YES', 0 => 'NO']], 'explain' => false),
```
Comment (Derky) [inline phpBB/includes/acp/acp_attachments.php:179]: I see you also got rid of the `radio:` shorthands for `yes_no` and `enabled_disabled`. Those are actually nice to be able to reuse. Is it possible to keep those? (as long as the HTML is generated in the Twig macro)

Comment (rxu) [reply phpBB/includes/acp/acp_attachments.php:179]: @Derky Done.

File: phpBB/phpbb/avatar/manager.php [hunk 6]
```diff
@@ -268,9 +268,16 @@ public function get_avatar_settings($driver)
 	{
 		$config_name = $driver->get_config_name();
 
-		return array(
-			'allow_avatar_' . $config_name	=> array('lang' => 'ALLOW_' . strtoupper(str_replace('\\', '_', $config_name)),		'validate' => 'bool',	'type' => 'radio:yes_no', 'explain' => true),
-		);
+		return [
+			'allow_avatar_' . $config_name	=> [
```
Comment (Derky) [inline phpBB/phpbb/avatar/manager.php:272]: Thanks for restoring the radio:yes_no and enabled_disabled options. Can you please revert this code to use the radio:yes_no again?

File: phpBB/includes/acp/acp_attachments.php [hunk 7]
```diff
@@ -176,26 +176,26 @@ function main($id, $mode)
...
 						'max_filesize'			=> array('lang' => 'ATTACH_MAX_FILESIZE',	'validate' => 'string',	'type' => 'custom', 'method' => 'max_filesize', 'explain' => true),
 						'max_filesize_pm'		=> array('lang' => 'ATTACH_MAX_PM_FILESIZE','validate' => 'string',	'type' => 'custom', 'method' => 'max_filesize', 'explain' => true),
-						'secure_downloads'		=> array('lang' => 'SECURE_DOWNLOADS',		'validate' => 'bool',	'type' => 'radio:yes_no', 'explain' => true),
-						'secure_allow_deny'		=> array('lang' => 'SECURE_ALLOW_DENY',		'validate' => 'int',	'type' => 'custom', 'method' => 'select_allow_deny', 'explain' => true),
-						'secure_allow_empty_referer'	=> array('lang' => 'SECURE_EMPTY_REFERRER', 'validate' => 'bool',	'type' => 'radio:yes_no', 'explain' => true),
-						'check_attachment_content' 		=> array('lang' => 'CHECK_CONTENT', 'validate' => 'bool',	'type' => 'radio:yes_no', 'explain' => true),
+						'secure_downloads'		=> array('lang' => 'SECURE_DOWNLOADS',		'validate' => 'bool',	'type' => 'radio', 'function' => 'build_radio', 'params' => ['{CONFIG_VALUE}', '{KEY}', [1 => 'YES', 0 => 'NO']], 'explain' => true),
```
Comment (Derky) [inline phpBB/includes/acp/acp_attachments.php:187]: Can you please undo the changes to these yes_no options as well? Thanks!

**Comment by Derky** [comment]
Thanks for updating the PR! I've also asked @marc1706 to take a look at the radio options to see if it's aligned with his recent changes to form elements. 

I've been testing this PR around a bit and the board settings page is working again now, but I noticed there are no `selected="selected"` attributes on both the "Default style" as "Guest style". So this causes the first option in the selectbox to show as selected. (So if you select a different style, it doesn't show that as selected next time you return to the page)

**Comment by rxu** [comment]
> there are no selected="selected" attributes on both the "Default style" as "Guest style".

At what page?

**Comment by rxu** [comment]
@Derky Probably fixed "selected" issue, could you please retest it again.

File: phpBB/includes/functions_acp.php [hunk 8]
```diff
@@ -229,35 +229,41 @@ function build_select($option_ary, $option_default = false): array
...
+ * @param array		$tpl_type	Configuration setting type data
+ * @param string	$key		Configuration option name
+ * @param array		$new_ary	Updated configuration data
+ * @param string	$config_key	Configuration option name
+ * @param Array		$vars		Configuration setting data
+ *
+ * @return array|string
  */
-function phpbb_build_cfg_template(array $tpl_type, string $key, &$new_ary, $config_key, $vars): array
+function build_cfg_template($tpl_type, $key, &$new_ary, $config_key, $vars)
```
Comment (marc1706) [inline phpBB/includes/functions_acp.php:264]: Looks like you merge both functions into one? The phpbb_ prefixed one is actually newer and we started to prefix functions a while ago since functions don't have namespaces. If you merge these two, keep the prefix. ;)

File: phpBB/includes/ucp/ucp_prefs.php [hunk 9]
```diff
@@ -205,14 +205,18 @@ function main($id, $mode)
...
+						'options'	=> style_select($data['user_style'], false, $styles_row)
+					],
 					'TIMEZONE_OPTIONS'	=> [
 						'tag'		=> 'select',
 						'name'		=> 'tz',
 						'options'	=> $timezone_select,
 					],
-					'S_CAN_HIDE_ONLINE'		=> ($auth->acl_get('u_hideonline')) ? true : false,
-					'S_SELECT_NOTIFY'		=> ($config['jab_enable'] && $user->data['user_jabber'] && @extension_loaded('xml')) ? true : false)
+					'S_CAN_HIDE_ONLINE'	=> ($auth->acl_get('u_hideonline')) ? true : false,
```
Comment (marc1706) [inline phpBB/includes/ucp/ucp_prefs.php:218]: That can be simplified to `'S_CAN_HIDE_ONLINE' => $auth->acl_get('u_hideonline'),`

File: phpBB/includes/ucp/ucp_prefs.php [hunk 10]
```diff
@@ -205,14 +205,18 @@ function main($id, $mode)
...
+					'S_SELECT_NOTIFY'	=> ($config['jab_enable'] && $user->data['user_jabber'] && @extension_loaded('xml')) ? true : false)
```
Comment (marc1706) [inline phpBB/includes/ucp/ucp_prefs.php:219]: Same here

File: phpBB/includes/acp/acp_board.php [hunk 11]
```diff
@@ -1034,30 +1088,65 @@ public function language_select(string $default = '', array $langdata = []): arr
...
+
+	/**
+	 * Wrapper function for style_select()
+	 *
+	 * @param string $default	Style ID to be selected in the dropdown list
+	 * @param bool $all			Flag indicating if all styles data including inactive should be fetched
+	 *
+	 * @return array
+	 */
+	public function style_select($default, $all): array
```
Comment (marc1706) [inline phpBB/includes/acp/acp_board.php:1102]: Should be prefixed with phpbb_

Comment (marc1706) [reply phpBB/includes/acp/acp_board.php:1102]: Also, use type hints

File: phpBB/includes/functions_acp.php [hunk 12]
```diff
@@ -229,35 +229,41 @@ function build_select($option_ary, $option_default = false): array
 /**
 * Build radio fields in acp pages
 */
-function h_radio($name, $input_ary, $input_default = false, $id = false, $key = false, $separator = '')
+function build_radio($value, $key, $options)
```
Comment (marc1706) [inline phpBB/includes/functions_acp.php:232]: Use phpbb_ prefix

**Review by marc1706 — CHANGES_REQUESTED** [review]

**Comment by Derky** [comment]
> @Derky Probably fixed "selected" issue, could you please retest it again.

Yes this was fixed, thanks!

**Comment by rxu** [comment]
@marc1706 Done.

File: phpBB/styles/all/template/macros/forms/textarea.twig
Comment (rubencm) [inline phpBB/styles/all/template/macros/forms/textarea.twig:1]: Please, can you make id optional?

File: phpBB/includes/functions_acp.php [hunk 13]
```diff
@@ -539,15 +505,15 @@ function build_cfg_template($tpl_type, $key, &$new_ary, $config_key, $vars)
...
+	* @var	array			tpl_type	Config type array:
+	*							0 => data type
+	*							1 [optional] => string: size, int: minimum
+	*							2 [optional] => string: max. length, int: maximum
+	* @var	string			key			Should be used for the id attribute in html
+	* @var	array			new			Array with the config values we display
+	* @var	string			name		Should be used for the name attribute
+	* @var	array			vars		Array with the options for the config
+	* @var	array|string	tpl			The resulting html code we display
 	* @since 3.1.0-a1
```
[the remark this answers is not in the document]
Comment (marc1706) [reply phpBB/includes/functions_acp.php:517]: Yes, use `@changed` for changes to events.

[the remark this answers is not in the document]
Comment (rxu) [reply phpBB/includes/functions_acp.php:517]: Just to make sure, shall I rename the event or just add an info that event location's function name was changed?

File: phpBB/includes/functions_acp.php [hunk 14]
```diff
@@ -368,154 +386,102 @@ function phpbb_build_cfg_template(array $tpl_type, string $key, &$new_ary, $conf
...
-		break;
-	}
-
-	return $tpl;
-}
-
-/**
-* Build configuration template for acp configuration pages
-*/
-function build_cfg_template($tpl_type, $key, &$new_ary, $config_key, $vars)
```
[the remark this answers is not in the document]
Comment (marc1706) [reply phpBB/includes/functions_acp.php:409]: I'd say here: https://github.com/phpbb/documentation/blob/master/development/development/upgrade/400.rst

File: phpBB/includes/functions_acp.php [hunk 15]
```diff
@@ -228,37 +233,50 @@ function build_select($option_ary, $option_default = false): array
...
+ * Build radio fields in acp pages
+ *
+ * @param int|string	$value	Configuration option value
+ * @param string		$key	Configuration option key name
+ * @param array			$options Configuration options data
+ * 							representing array of [values => language_keys]
+ *
+ * @return array
+ */
+function phpbb_build_radio(int|string $value, string $key, array $options): array
```
[the remark this answers is not in the document]
Comment (marc1706) [reply phpBB/includes/functions_acp.php:245]: Here: https://github.com/phpbb/documentation/blob/master/development/development/upgrade/400.rst

File: phpBB/includes/functions_acp.php (hunk 14 above)
[the remark this answers is not in the document]
Comment (rxu) [reply phpBB/includes/functions_acp.php:409]: https://github.com/phpbb/documentation/pull/289

File: phpBB/includes/functions_acp.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (rxu) [reply phpBB/includes/functions_acp.php:245]: https://github.com/phpbb/documentation/pull/289

**Review by marc1706 — APPROVED** [review]
